src/relay/backend/te_compiler.cc

```diff
@@ -952,25 +919,24 @@ backend::FunctionInfo UpdateMainWorkspaceSize(const IRModule& mod, tec::TargetMa
   Map<Target, Function> relay_primfuncs;
 
   // Initialize all target workspaces to zero
-  for (const auto& kv : targets) {
-    auto tgt = kv.second;
-    workspace_sizes.Set(tgt, 0);
+  for (const auto& target : config->primitive_targets) {
+    workspace_sizes.Set(target, 0);
   }
 
   for (const auto& dev_and_size : device_workspace) {
-    auto tgt = tec::GetTargetFromInteger(dev_and_size.first, targets);
-    workspace_sizes.Set(tgt, dev_and_size.second);
-    relay_primfuncs.Set(tgt, func);
+    Target target = config->FindPrimitiveTargetOrFail(dev_and_size.first);
```

Comment:
It looks like the main use of `FindPrimitiveTargetOrFail` is to de-duplicate targets that share resource pools.  What is the advantage of using the target itself to handle this de-duplication, rather than using `target->kind->device_type` as the key for tracking resource usage?

Reply:
Indeed we are still in a transitionary state, moving from using device types as the only distinguishing property in the heterogenous case to using for VirtualDevices. Eventually FindPrimitiveTargetOrFail will only be needed by the PlanDevices pass when it needs to choose targets to match its device assignment.

Reply:
(@Mousius  is continuing to chip away at the 3-way disconnect between "Compiler" targets (strings), Targets and device types as the carrier of target information.)

Reply:
Got it!  So after the transitionary state, rather than needing to assume that all targets that use a specific `device_type` will share the same resource pool, the equivalent deduplication would be implicit in the virtual device assignment?

Reply:
That was the plan as of last year, we just didn't get to finish the programme.